eventio/event_io_file.py

```diff
@@ -2,52 +2,64 @@
 from .object_header import make_ObjectHeader
 from io import BytesIO
 
-__all__ = ['object_headers', 'yield_objects']
+class EventIOObject:
+    ''' A generic EventIOObject
+
+    It has a list of `headers` and a binary string `payload`.
+    The payload might be loaded lazily on first access or already in memory.
+    This is decided on construction time.
+    '''
+    def __init__(self, headers, payload=None, file=None):
+        if file is not None:
+            self._file = file
+        elif payload is not None:
+            self._file = None
+            self.payload = payload
+        else:
+            raise Exception('EventIOObject must have either file or payload')
         
-def read_all_object_headers(f, toplevel=True):
-    '''f is something like io.BufferedIOBase'''
-    object_headers = []
-    while True:
-        try:
-            header = make_ObjectHeader(f, toplevel)
-            payload = f.read(header.length)
-            if not header.only_sub_objects:
-                object_headers.append(header)
-            else:
-                sub_object_headers = read_all_object_headers(BytesIO(payload), toplevel=False)
-                object_headers.append((header, sub_object_headers))
+        self.headers = headers
 
-        except ValueError:
-            warnings.warn('File seems to be truncated')
-            break
-        except struct.error:
-            break
-    return object_headers
+    def __getattr__(self, attr):
+        if attr == "payload":
+            if self._file.closed:
+                self._file = open(self._file.name, 'rb')
+            start_address = sum(h.data_field_first_byte for h in self.headers)
+            self._file.seek(start_address)
+            self.payload = self._file.read(self.headers[-1].length)
+        return self.payload
 
+    def __repr__(self):
+        return repr(self.headers)
 
-def object_headers(path):
+def object_generator(path):
     with open(path, 'rb') as f:
-        return read_all_object_headers(f)
+        for o in yield_all_objects(f, read_payload=True):
+            yield o
+
+def object_list(path):
+    f = open(path, 'rb')
```

Comment:
I think you should give it a file object. Like this we keep compatibility with all supported compression formats of python (gzip, bzip2 etc)